drivers/dma/dma_cavs_gpdma.c

```diff
@@ -0,0 +1,137 @@
+
+#define GPDMA_CTL_OFFSET 0x004
+#define GPDMA_CTL_FDCGB BIT(0)
+
+#include "dma_dw_common.h"
+
+#define LOG_LEVEL CONFIG_DMA_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(dma_cavs_gpdma);
+
+
+/* Device run time data */
+struct cavs_gpdma_data {
+	struct dw_dma_dev_data dw_data;
+};
+
+/* Device constant configuration parameters */
+struct cavs_gpdma_cfg {
+	struct dw_dma_dev_cfg dw_cfg;
+	uint32_t shim;
+};
+
+/* Disables automatic clock gating (force disable clock gate) */
+static void cavs_gpdma_clock_enable(const struct device *dev)
+{
+	const struct cavs_gpdma_cfg *const dev_cfg = dev->config;
+	uint32_t reg = dev_cfg->shim + GPDMA_CTL_OFFSET;
+
+	sys_write32(GPDMA_CTL_FDCGB, reg);
+}
+
+
+int cavs_gpdma_init(const struct device *dev)
+{
+	const struct cavs_gpdma_cfg *const dev_cfg = dev->config;
+
+	/* Disable dynamic clock gating appropriately before initializing */
+	cavs_gpdma_clock_enable(dev);
+
+	/* Disable all channels and Channel interrupts */
+	dw_dma_setup(dev);
+
+	/* Configure interrupts */
+	dev_cfg->dw_cfg.irq_config();
+
+	LOG_INF("Device %s initialized", dev->name);
+
+	return 0;
+}
+
+
+static const struct dma_driver_api cavs_gpdma_driver_api = {
+	.config = dw_dma_config,
+	.reload = dw_dma_reload,
+	.start = dw_dma_transfer_start,
+	.stop = dw_dma_transfer_stop,
+};
+
+#define CAVS_GPDMA_INIT(inst)						\
+									\
+	static struct dw_drv_plat_data dmac##inst = {			\
+		.chan[0] = {						\
+			.class  = 6,					\
+			.weight = 0,					\
+		},							\
+		.chan[1] = {						\
+			.class  = 6,					\
+			.weight = 0,					\
+		},							\
+		.chan[2] = {						\
+			.class  = 6,					\
+			.weight = 0,					\
+		},							\
+		.chan[3] = {						\
+			.class  = 6,					\
+			.weight = 0,					\
+		},							\
+		.chan[4] = {						\
+			.class  = 6,					\
+			.weight = 0,					\
+		},							\
+		.chan[5] = {						\
+			.class  = 6,					\
+			.weight = 0,					\
+		},							\
+		.chan[6] = {						\
+			.class  = 6,					\
+			.weight = 0,					\
+		},							\
+		.chan[7] = {						\
+			.class  = 6,					\
+			.weight = 0,					\
+		},							\
+	};								\
+									\
+	static void cavs_gpdma##inst##_irq_config(void);			\
+									\
+	static const struct cavs_gpdma_cfg cavs_gpdma##inst##_config = { \
+		.dw_cfg = {						\
+			.base = DT_INST_REG_ADDR(inst),			\
+			.irq_config = cavs_gpdma##inst##_irq_config,	\
+		},							\
+		.shim = DT_INST_PROP_BY_IDX(inst, shim, 0),		\
+	};								\
+									\
+	static struct cavs_gpdma_data cavs_gpdma##inst##_data = {	\
+		.dw_data = {						\
+			.channel_data = &dmac##inst,			\
+		},							\
+	};								\
+									\
+									\
+	DEVICE_DT_INST_DEFINE(inst,					\
+			      &cavs_gpdma_init,				\
+			      NULL,					\
+			      &cavs_gpdma##inst##_data,			\
+			      &cavs_gpdma##inst##_config, POST_KERNEL,	\
+			      CONFIG_DMA_INIT_PRIORITY,			\
+			      &cavs_gpdma_driver_api);			\
+									\
+	static void cavs_gpdma##inst##_irq_config(void)			\
```

Comment:
Why not just put the irq into the config struct as a field and use a single function?  The IRQ_CONNECTS() for all instances can then be put into a much smaller FOREACH in a global sysinit or something.  (I mean, I agree that this is a spot where IRQ_CONNECT()'s inability to be declared statically sort of bites us, but I'm not sure defining a separate nearly-noop function for every instance is the cleanest workaround.)

Reply:
IRQ_CONNECT boils down to, as I recall, a table in static memory in a specified linker section for xtensa. I can't use a struct variable, even if const and static itself (its not a const expression for the macro expansion) to dynamically do IRQ_CONNECT(). I could use irq_connect_dynamic with a common function and save only the irq num here. It's not necessarily the common thing to do as on other platforms IRQ_CONNECT puts things in flash I believe to save memory on the vector table. Does that matter here? probably not right.

Reply:
I misunderstood the point here last week. I get it now, but to be fair this is the way the dma dw driver was written previously. I could do the above but I actually prefer to see the CONNECT and enable colocated here as it makes it easy to understand. I could do what I initially thought you meant, and save the irq num and swap to irq_connect_dynamic here with a single function but I'm fine with the way it is.

It would be nice if IRQ_CONNECT could be statically declared, and looking at the macros again it seems like it really could be statically declared for Xtensa at least. I pushed a commit that does just that and uses it here. This commit could be dropped if its not liked.